contrib/node/src/python/pants/contrib/node/subsystems/package_managers.py

```diff
@@ -0,0 +1,231 @@
+  PROD = 'prod'
+  DEV = 'dev'
+  PEER = 'peer'
+  BUNDLE = 'bundle'
+  OPTIONAL = 'optional'
+  NO_SAVE = 'not saved'
+
+
+class PackageInstallationVersionOption(object):
+  EXACT = 'exact'
+  TILDE = 'tilde'
+
+
+class PackageManager(object):
+  """Defines node package manager functionalities."""
+
+  def __init__(self, name, tool_installations):
+    self.name = name
+    self.tool_installations = tool_installations
+
+  def _get_installation_args(self, install_optional, production_only, force):
+    """Returns command line args for installing package.
+
+    :param install_optional: True to request install optional dependencies.
+    :param production_only: True to only install production dependencies, i.e.
+      ignore devDependencies.
+    :param force: True to force re-download dependencies.
+    :rtype: list of strings
+    """
+    raise NotImplementedError
+
+  def _get_run_script_args(self):
+    """Returns command line args to run a package.json script.
+
+    :rtype: list of strings
+    """
+    raise NotImplementedError
+
+  def _get_add_package_args(self, package, type_option, version_option):
+    """Returns command line args to add a node pacakge.
+
+    :rtype: list of strings
+    """
+    raise NotImplementedError()
+
+  def run_command(self, args=None, node_paths=None):
+    """Returns a command that when executed will run an arbitury command via package manager."""
+    return command_gen(
+      self.tool_installations,
+      self.name,
+      args=args,
+      node_paths=node_paths
+    )
+
+  def install_module(
+    self,
+    install_optional=False,
+    production_only=False,
+    force=False,
+    node_paths=None):
+    """Returns a command that when executed will install node package.
+
+    :param install_optional: True to install optional dependencies.
+    :param production_only: True to only install production dependencies, i.e.
+      ignore devDependencies.
+    :param force: True to force re-download dependencies.
+    :param node_paths: A list of path that should be included in $PATH when
+      running installation.
+    """
+    args=self._get_installation_args(
+      install_optional=install_optional,
+      production_only=production_only,
+      force=force)
+    return self.run_command(args=args, node_paths=node_paths)
+
+  def run_script(self, script_name, script_args=None, node_paths=None):
+    """Returns a command to execute a package.json script.
+
+    :param script_name: Name of the script to name.  Note that script name 'test'
+      can be used to run node tests.
+    :param script_args: Args to be passed to package.json script.
+    :param node_paths: A list of path that should be included in $PATH when
+      running the script.
+    """
+    package_manager_args = self._get_run_script_args()
+    package_manager_args.append(script_name)
+    if script_args:
+      package_manager_args.append('--')
+      package_manager_args.extend(script_args)
```

Comment:
Even adding arguments like this could be useful to add to `util/` at some point. I feel like having some common framework to share manipulation of argument lists for invoking pants-owned binaries would be great -- a method adding `'--'` and the positional args could be used to wrap invocations across many different binaries without tedious error-prone argument list creation. You can see where this would have helped with something like #4958.

Just a thought -- no action needed here in this PR.

Reply:
Maybe just a TODO to consider it...

Reply:
done.